libraries/AP_TTLServo/AP_TTLServo.cpp

```diff
@@ -0,0 +1,418 @@
+{
+    uint8_t id = packet[PKT_ID];
+
+    // Discard servos beyond the maximum permissible number of servo channels
+    if (id > NUM_SERVO_CHANNELS) {
+        return;
+    }
+
+    // If the servo wasn't previously identified, mark its existence on the network
+    uint32_t id_mask = (1U<<(id));
+    if (!(id_mask & servo_id_mask)) {
+        servo_id_mask |= id_mask;
+        hal.console->printf("TTL servo with ID %u identified\n", id);
+    }
+}
+
+// Read the bytes received from responses
+void AP_TTLServo::read_bytes(void)
+{
+    uint32_t n = port->available();
+    // If no bytes received or received less than the required to decode an
+    // instruction, return in order to wait for the required number of bytes
+    if (n == 0 && pktbuf_ofs < PKT_INSTRUCTION) {
+        return;
+    }
+
+    // Read from serial the maximum number of bytes that would fill the buffer
+    if (n > sizeof(pktbuf) - pktbuf_ofs) {
+        n = sizeof(pktbuf) - pktbuf_ofs;
+    }
+    for (uint8_t i = 0; i < n; i++) {
+        pktbuf[pktbuf_ofs++] = port->read();
+    }
+
+    // Discard bad leading data. This should be rare
+    while (pktbuf_ofs >= 2 && (pktbuf[0] != 0xFF || pktbuf[1] != 0xFF)) {
+        memmove(pktbuf, &pktbuf[1], pktbuf_ofs-1);
+        pktbuf_ofs--;
+    }
+
+    // If enough data hasn't been received, return
+    if (pktbuf_ofs < 5) {
+        return;
+    }
+
+    // Check if enough data has been received according to the Packet.
+    // If it hasn't been received, return and wait for the rest of the Packet
+    const uint8_t total_packet_length = pktbuf[PKT_LENGTH] + PKT_INSTRUCTION;
+    if (total_packet_length > sizeof(pktbuf)) {
+        pktbuf_ofs = 0;
+        return;
+    }
+    if (pktbuf_ofs < total_packet_length) {
+        return;
+    }
+
+    // Compare the Packet's CRC with the received Packet data. If it is equal,
+    // the Packet has been received without data errors. Otherwise, just discard
+    // the received Packet (had errors)
+    const uint8_t crc = pktbuf[total_packet_length - 1];
+    const uint8_t calc_crc = calculate_crc(pktbuf, total_packet_length - 1);
+    if (calc_crc == crc) {
+      // Process full packet
+      process_packet(pktbuf, total_packet_length);
+    }
+
+    // Removed the processed Packet data from the buffer
+    memmove(pktbuf, &pktbuf[total_packet_length], pktbuf_ofs - total_packet_length);
+    pktbuf_ofs -= total_packet_length;
+}
+
+// Send a command to the servos, changing a register value
+void AP_TTLServo::send_command(uint8_t id, uint8_t reg, uint16_t value, uint8_t len)
+{
+    struct packet {
+        uint8_t id = id;                       //#1 Servo ID
+        uint8_t length;                   //#2 Packet length 
+        uint8_t instruction = INST_WRITE; //#3 Instruction is a Write
+        uint8_t reg;                      //#4 First parameter is the register
+        uint16_t value;           //#5 Following parameters is the value
+    } tx_packet;
```

Comment:
@teoPiD 

```suggestion
void AP_TTLServo::send_command(uint8_t val, uint8_t reg, uint16_t value, uint8_t len)
{
    struct packet {
        uint8_t id = val;                       //#1 Servo ID
        uint8_t length;                   //#2 Packet length 
        uint8_t instruction = INST_WRITE; //#3 Instruction is a Write
        uint8_t reg;                      //#4 First parameter is the register
        uint16_t value;           //#5 Following parameters is the value
    } tx_packet;
```

Reply:
Here is how I would have done this. *not review of your code just this bit

```suggestion
void AP_TTLServo::send_command(uint8_t id, uint8_t reg, uint16_t value, uint8_t len)
{
    struct packet {
        uint8_t id;             //#1 Servo ID
        uint8_t length;         //#2 Packet length 
        uint8_t instruction;    //#3 Instruction is a Write
        uint8_t reg;            //#4 First parameter is the register
        uint16_t value;         //#5 Following parameters is the value
    } tx_packet =
        { val, uint8_t(3 + len), INST_WRITE, reg, value };
```